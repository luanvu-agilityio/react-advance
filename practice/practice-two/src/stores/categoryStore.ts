import productApi from '@services/product'
import { create } from 'zustand'
import { devtools, subscribeWithSelector } from 'zustand/middleware'

interface CategoryState {
  // Pagination
  currentPage: number
  limit: number
  displayLimit: number

  // Filters
  category?: string
  subcategory?: string
  selectedBrands: string[]
  selectedRatings: number[]
  priceRange: { min: number; max: number }
  searchQuery?: string

  // UI State
  viewMode: string

  //Sorting
  sortBy: string
  sortOrder: 'asc' | 'desc'
  setSortBy: (field: string) => void
  setSortOrder: (order: 'asc' | 'desc') => void

  // Actions
  setPage: (page: number) => void
  setLimit: (limit: number) => void
  setCategory: (category?: string) => void
  setSubcategory: (subcategory?: string) => void
  setBrands: (brands: string[]) => void
  setRatings: (ratings: number[]) => void
  setPriceRange: (range: { min: number; max: number }) => void
  setSearchQuery: (query?: string) => void
  setViewMode: (mode: string) => void
  resetFilters: () => void
  resetPagination: () => void
  setDisplayLimit: (limit: number) => void

  // URL helpers
  getUrlParams: () => URLSearchParams
  setFromUrl: (searchParams: URLSearchParams) => void
  getApiParams: () => Record<string, string | number | string[] | number[]>
  getProductCountBySubcategory: (subcategory: string) => Promise<number>
}

const defaultPriceRange = { min: 0, max: 1000 }

export const useCategoryStore = create<CategoryState>()(
  devtools(
    subscribeWithSelector((set, get) => ({
      // Initial state
      currentPage: 1,
      limit: 5,
      displayLimit: 5,
      selectedBrands: [],
      selectedRatings: [],
      priceRange: defaultPriceRange,
      viewMode: 'grid',
      sortBy: 'name',
      sortOrder: 'asc',

      // Sorting actions
      // Sort actions
      setSortBy: (field: string) => {
        set({ sortBy: field, currentPage: 1 })
      },

      setSortOrder: (order: 'asc' | 'desc') => {
        set({ sortOrder: order, currentPage: 1 })
      },

      // Pagination actions
      setPage: (page: number) => {
        set({ currentPage: Math.max(1, page) })
      },

      setLimit: (limit: number) => {
        set({ limit: Math.max(1, limit), currentPage: 1 })
      },

      // Filter actions
      setCategory: (category) => {
        set({ category, currentPage: 1 }) // Reset page when category changes
      },

      setSubcategory: (subcategory) => {
        set({ subcategory, currentPage: 1 })
      },

      setBrands: (brands) => {
        set({ selectedBrands: brands, currentPage: 1 })
      },

      setRatings: (ratings) => {
        set({ selectedRatings: ratings, currentPage: 1 })
      },

      setPriceRange: (range) => {
        set({ priceRange: range, currentPage: 1 })
      },

      setSearchQuery: (query) => {
        set({ searchQuery: query, currentPage: 1 })
      },

      setViewMode: (mode) => {
        set({ viewMode: mode })
      },

      setDisplayLimit: (displayLimit: number) => {
        const newLimit = Math.max(1, displayLimit)
        set({
          displayLimit: newLimit,
          limit: newLimit,
          currentPage: 1,
        })
      },
      // Reset functions
      resetFilters: () => {
        set({
          selectedBrands: [],
          selectedRatings: [],
          priceRange: defaultPriceRange,
          subcategory: undefined,
          currentPage: 1,
        })
      },

      resetPagination: () => {
        set({ currentPage: 1 })
      },

      // URL management
      getUrlParams: () => {
        const state = get()
        const params = new URLSearchParams()

        if (state.currentPage > 1) {
          params.set('page', state.currentPage.toString())
        }

        // Always include limit in URL params
        params.set('limit', state.limit.toString())

        if (state.subcategory) {
          params.set(
            'subcategory',
            state.subcategory.toLowerCase().replace(/\s+/g, '-')
          )
        }

        if (state.searchQuery) {
          params.set('search', state.searchQuery)
        }

        params.set('sort', state.sortBy)
        params.set('order', state.sortOrder)

        return params
      },

      setFromUrl: (searchParams) => {
        const page = parseInt(searchParams.get('page') || '1', 10)
        const limit = parseInt(searchParams.get('limit') || '5', 10)
        const subcategory = searchParams.get('subcategory')
        const search = searchParams.get('search')
        const sort = searchParams.get('sort') || 'name'
        const order = searchParams.get('order') === 'desc' ? 'desc' : 'asc'

        set({
          currentPage: Math.max(1, page),
          limit: Math.max(1, limit),
          subcategory: subcategory || undefined,
          searchQuery: search || undefined,
          sortBy: sort,
          sortOrder: order as 'asc' | 'desc',
        })
      },

      // API params
      getApiParams: () => {
        const state = get()
        const params: Record<string, string | number | string[] | number[]> = {
          p: state.currentPage,
          l: state.limit,
        }
        if (state.category) params.category = state.category
        if (state.subcategory) params.subcategory = state.subcategory
        if (state.selectedBrands.length > 0)
          params.brands = state.selectedBrands
        if (state.priceRange.min > defaultPriceRange.min)
          params.minPrice = state.priceRange.min
        if (state.priceRange.max < defaultPriceRange.max)
          params.maxPrice = state.priceRange.max
        if (state.searchQuery) params.search = state.searchQuery
        if (state.selectedRatings.length > 0)
          params.ratings = state.selectedRatings

        params.sortBy = state.sortBy
        params.sortOrder = state.sortOrder
        return params
      },

      getProductCountBySubcategory: async (subcategory: string) => {
        try {
          // Make a specific API call to count products
          const response = await productApi.getProductCount({ subcategory })
          return response
        } catch (error) {
          console.error('Error getting product count:', error)
          return 0
        }
      },
    })),
    { name: 'category-store' }
  )
)
