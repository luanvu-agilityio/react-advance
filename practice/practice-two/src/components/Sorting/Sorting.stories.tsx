import type { Meta, StoryObj } from '@storybook/react'
import { Sorting } from './Sorting'
import { useCategoryStore } from '@stores/categoryStore'
import { useEffect } from 'react'

// Create a wrapper that sets up the store with initial values
const SortingWithStore = ({
  initialSortBy = 'name',
  initialSortOrder = 'asc' as 'asc' | 'desc',
}: {
  initialSortBy?: string
  initialSortOrder?: 'asc' | 'desc'
}) => {
  const { setSortBy, setSortOrder } = useCategoryStore()

  // Set the initial values when the story loads
  useEffect(() => {
    setSortBy(initialSortBy)
    setSortOrder(initialSortOrder)
  }, [initialSortBy, initialSortOrder, setSortBy, setSortOrder])

  return <Sorting />
}

const meta: Meta<typeof Sorting> = {
  title: 'Components/Sorting',
  component: Sorting,
  parameters: {
    layout: 'centered',
    // Reset store between stories
    zustand: { resetStores: true },
  },
  tags: ['autodocs'],
  decorators: [
    (Story) => (
      <div
        style={{
          padding: '20px',
          backgroundColor: '#f9f9f9',
          borderRadius: '8px',
          minWidth: '300px',
        }}
      >
        <Story />
      </div>
    ),
  ],
}

export default meta
type Story = StoryObj<typeof Sorting>

// Default sort by name, ascending
export const Default: Story = {
  render: () => (
    <SortingWithStore initialSortBy="name" initialSortOrder="asc" />
  ),
}

// Sort by price, ascending
export const SortByPrice: Story = {
  render: () => (
    <SortingWithStore initialSortBy="price" initialSortOrder="asc" />
  ),
}

// Sort by rating, ascending
export const SortByRating: Story = {
  render: () => (
    <SortingWithStore initialSortBy="rating" initialSortOrder="asc" />
  ),
}

// Sort by name, descending
export const DescendingOrder: Story = {
  render: () => (
    <SortingWithStore initialSortBy="name" initialSortOrder="desc" />
  ),
}
