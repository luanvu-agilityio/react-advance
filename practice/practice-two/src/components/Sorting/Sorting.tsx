import { useCategoryStore } from '@stores/categoryStore'
import Select from '@components/common/Select'
import { SelectWrapper } from '@components/Searchbar/Searchbar.styles'
import {
  FilterControlContainer,
  FilterLabel,
  FilterRadio,
  FilterValue,
} from '@components/Pagination/PaginationStyles'

const sortOptions = [
  { value: 'name', label: 'Name' },
  { value: 'price', label: 'Price' },
  { value: 'rating', label: 'Rating' },
]

export const Sorting = () => {
  const { sortBy, sortOrder, setSortBy, setSortOrder } = useCategoryStore()

  return (
    <FilterControlContainer style={{ minWidth: '180px' }}>
      <FilterLabel>Sort By :</FilterLabel>
      <SelectWrapper
        style={{ width: 'auto', minWidth: '100px', margin: '0', padding: 0 }}
      >
        <Select
          options={sortOptions}
          value={sortBy}
          onChange={(value) => setSortBy(value)}
          variant="filter"
        />
      </SelectWrapper>
      <div style={{ display: 'flex', gap: '12px', alignItems: 'center' }}>
        <label
          style={{
            display: 'flex',
            alignItems: 'center',
            gap: '4px',
            cursor: 'pointer',
          }}
        >
          <FilterRadio
            name="sortOrder"
            checked={sortOrder === 'asc'}
            onChange={() => setSortOrder('asc')}
          />
          <FilterValue $isActive={sortOrder === 'asc'}>
            <span>Ascending</span>
          </FilterValue>
        </label>
        <label
          style={{
            display: 'flex',
            alignItems: 'center',
            gap: '4px',
            cursor: 'pointer',
          }}
        >
          <FilterRadio
            name="sortOrder"
            checked={sortOrder === 'desc'}
            onChange={() => setSortOrder('desc')}
          />
          <FilterValue $isActive={sortOrder === 'desc'}>
            <span>Descending</span>
          </FilterValue>
        </label>
      </div>
    </FilterControlContainer>
  )
}
