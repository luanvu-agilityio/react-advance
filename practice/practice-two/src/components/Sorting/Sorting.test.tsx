import { render, screen, fireEvent } from '@testing-library/react'
import { Sorting } from './Sorting'
import { useCategoryStore } from '@stores/categoryStore'

// Mock the Select component
jest.mock('@components/common/Select', () => {
  return {
    __esModule: true,
    default: jest.fn(({ onChange, value }) => (
      <div data-testid="select-mock">
        <select
          value={value}
          data-testid="mock-select-element"
          onChange={(e) => onChange(e.target.value)}
        >
          <option value="name">Name</option>
          <option value="price">Price</option>
          <option value="rating">Rating</option>
        </select>
      </div>
    )),
  }
})

// Mock the Zustand store
jest.mock('@stores/categoryStore', () => ({
  useCategoryStore: jest.fn(),
}))

describe('Sorting', () => {
  // Mock store functions
  const mockSetSortBy = jest.fn()
  const mockSetSortOrder = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()

    // Default mock implementation
    ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
      sortBy: 'name',
      sortOrder: 'asc',
      setSortBy: mockSetSortBy,
      setSortOrder: mockSetSortOrder,
    })
  })

  // Snapshot tests
  describe('rendering', () => {
    test('renders correctly with default values', () => {
      const { asFragment } = render(<Sorting />)
      expect(asFragment()).toMatchSnapshot()
    })

    test('renders correctly with price sorting', () => {
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        sortBy: 'price',
        sortOrder: 'asc',
        setSortBy: mockSetSortBy,
        setSortOrder: mockSetSortOrder,
      })

      const { asFragment } = render(<Sorting />)
      expect(asFragment()).toMatchSnapshot()
    })

    test('renders correctly with descending order', () => {
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        sortBy: 'name',
        sortOrder: 'desc',
        setSortBy: mockSetSortBy,
        setSortOrder: mockSetSortOrder,
      })

      const { asFragment } = render(<Sorting />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  // Interaction tests
  describe('interactions', () => {
    test('calls setSortBy when dropdown value changes', () => {
      render(<Sorting />)

      // Find the select element in our mock
      const selectElement = screen.getByTestId('mock-select-element')

      // Trigger change event
      fireEvent.change(selectElement, { target: { value: 'price' } })

      expect(mockSetSortBy).toHaveBeenCalledWith('price')
    })

    test('calls setSortOrder with "desc" when descending option is clicked', () => {
      render(<Sorting />)
      const descRadio = screen.getByLabelText('Descending')

      fireEvent.click(descRadio)

      expect(mockSetSortOrder).toHaveBeenCalledWith('desc')
    })

    test('calls setSortOrder with "asc" when ascending option is clicked', () => {
      // Setup with desc initial value
      ;(useCategoryStore as unknown as jest.Mock).mockReturnValue({
        sortBy: 'name',
        sortOrder: 'desc',
        setSortBy: mockSetSortBy,
        setSortOrder: mockSetSortOrder,
      })

      render(<Sorting />)
      const ascRadio = screen.getByLabelText('Ascending')

      fireEvent.click(ascRadio)

      expect(mockSetSortOrder).toHaveBeenCalledWith('asc')
    })
  })
})
