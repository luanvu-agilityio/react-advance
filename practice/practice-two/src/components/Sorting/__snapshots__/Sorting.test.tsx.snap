// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting rendering renders correctly with default values 1`] = `
<DocumentFragment>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa bdMoJO"
    >
      Sort By :
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="select-mock"
      >
        <select
          data-testid="mock-select-element"
        >
          <option
            value="name"
          >
            Name
          </option>
          <option
            value="price"
          >
            Price
          </option>
          <option
            value="rating"
          >
            Rating
          </option>
        </select>
      </div>
    </div>
    <div
      style="display: flex; gap: 12px; align-items: center;"
    >
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          checked=""
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ cNZMBK"
        >
          <span>
            Ascending
          </span>
        </span>
      </label>
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ jRwoUa"
        >
          <span>
            Descending
          </span>
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sorting rendering renders correctly with descending order 1`] = `
<DocumentFragment>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa bdMoJO"
    >
      Sort By :
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="select-mock"
      >
        <select
          data-testid="mock-select-element"
        >
          <option
            value="name"
          >
            Name
          </option>
          <option
            value="price"
          >
            Price
          </option>
          <option
            value="rating"
          >
            Rating
          </option>
        </select>
      </div>
    </div>
    <div
      style="display: flex; gap: 12px; align-items: center;"
    >
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ jRwoUa"
        >
          <span>
            Ascending
          </span>
        </span>
      </label>
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          checked=""
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ cNZMBK"
        >
          <span>
            Descending
          </span>
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sorting rendering renders correctly with price sorting 1`] = `
<DocumentFragment>
  <div
    class="sc-fpEFIB ksmUmw"
    style="min-width: 180px; height: 60px;"
  >
    <span
      class="sc-gAqISa bdMoJO"
    >
      Sort By :
    </span>
    <div
      class="sc-fWnslK jxddlX"
      style="width: auto; min-width: 100px; margin: 0px; padding: 0px;"
    >
      <div
        data-testid="select-mock"
      >
        <select
          data-testid="mock-select-element"
        >
          <option
            value="name"
          >
            Name
          </option>
          <option
            value="price"
          >
            Price
          </option>
          <option
            value="rating"
          >
            Rating
          </option>
        </select>
      </div>
    </div>
    <div
      style="display: flex; gap: 12px; align-items: center;"
    >
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          checked=""
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ cNZMBK"
        >
          <span>
            Ascending
          </span>
        </span>
      </label>
      <label
        style="display: flex; align-items: center; gap: 4px; cursor: pointer;"
      >
        <input
          class="sc-fbguzk biTuMg"
          name="sortOrder"
          type="radio"
        />
        <span
          class="sc-brzPDJ jRwoUa"
        >
          <span>
            Descending
          </span>
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
